### PR TITLE
major cleanup.

### PR DESCRIPTION
**Pull request by DonutEspresso**
@yunong @micahr 

I think this is 'complete' now from my perspective, so would appreciate some feedback. The Error constructor signatures have been unified, and I have a restify branch integrated with all unit tests working. 

Pretty much everything is backwards compatible, except for one change. Errors sent to `res.send()` render with a `code` value that is basically the name of the error minus the word 'Error',  i.e., if you call:

```js
res.send(new errors.BadDigestError('indigestion));
// response renders as:
{
    code: 'BadDigest',
    message: 'indigestion'
}
```

There's an 'InternalError' (different from InternalServerError), which is used in the uncaught exception scenarios. When those errors are rendered out, you get:

```js
foo.bar();  // bad code, server will blow up
// response renders as :
{
    code: 'Internal',
    message: ''
}
```

Which can seem weird, because 'Internal' doesn't have a whole lot of meaning. There was some logic in the old impl which would strip/add the 'Error' suffix selectively for the `code` property depending on if it was InternalError or not, but this seems too one off case. I'd prefer we just rename 'Internal' to something else like 'InternalUncaught' if we need a better name, and/or to differentiate from InternalServerError.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -5,43 +5,64 @@ var util       = require('util');
...
  * @function makeError
- * @param    {String} name   the name of the error class to create
- * @param    {Number} [code] optional status code
- * @returns  {Object}
+ * @param    {String}   name       the name of the error class to create
+ * @param    {Number}   statusCode optional status code
+ * @returns  {Function}            a constructor function
  */
-function makeError(name, code) {
+function make(name, statusCode) {
```
Comment (micahr) [inline lib/index.js]: A little bit of an overloaded name. Perhaps something a little more descriptive?

File: lib/index.js [hunk 2]
```diff
@@ -5,43 +5,64 @@ var util       = require('util');
...
+    // pass every other arg down to constructor
+    return helpers.makeInstance(ErrCtor, args);
+}
+
 
 /**
  * create RestError subclasses for users. takes a string, creates a
  * constructor for them. magicks, again.
  * @public
  * @function makeError
```
Comment (micahr) [inline lib/index.js]: comment should line up with the function name.

File: tools/tasks/helpers.js [hunk 3]
```diff
@@ -68,15 +80,24 @@ function spawnBinary(options, cb) {
...
-    childProc.stderr.pipe(process.stderr);
+function spawnNodeBinary(options, cb) {
+    // assert required things are here
+    assert.object(options, 'options');
+    assert.arrayOfString(options.cmd, 'options.cmd');
+    assert.func(cb, 'callback');
+    assert.optionalString(options.errorMessage, 'options.errorMessage');
+
+    // prepend bin path with node_modules
+    options.cmd[0] = 'node_modules/.bin/' + options.cmd[0];
```
Comment (micahr) [inline tools/tasks/helpers.js]: Should probably be a constant.

File: lib/index.js (hunk 1 above)
Comment (DonutEspresso) [inline lib/index.js]: `makeConstructor` ?

**Comment by yunong** [comment]
Looks great, thanks for all the work. Taking a closer look now -- but please use make instead of gulp.

File: lib/baseClasses/HttpError.js [hunk 4]
```diff
@@ -22,21 +22,26 @@ function HttpError() {
...
+    // will handle calling captureStackTrace, using arguments.callee to
+    // eliminate unnecessary stack frames.
     WError.apply(self, args);
 
     /**
      * the error message.
      * when message isn't passed in with options, expect WError constructor
      * to fill in the message for us with the passed through args,
-     * usually via sprintf or just straight strings.
+     * usually via sprintf or just straight strings. thus, only set it
```
Comment (yunong) [inline lib/baseClasses/HttpError.js]: This seems like a bug with verror right? where if we set a message that's already been set, verror blows up? seems like a bug we should file with verror.

File: lib/baseClasses/HttpError.js [hunk 5]
```diff
@@ -61,12 +66,6 @@ function HttpError() {
         code: self.name,
         message: self.message || ''
     };
-
-    // capture the stack trace because that doesn't happen automatically.
-    // the second parameter into captureStackTrace() is used to filter out
-    // parts of the stack trace that aren't relevant to the users.
-    // i.e., the error constructors shouldn't show up in the stack trace.
-    Error.captureStackTrace(self, options.constructorOpt || self.constructor);
```
Comment (yunong) [inline lib/baseClasses/HttpError.js]: I must have missed this somewhere, but why aren't we capturing the stack trace anymore?

**Comment by yunong** [comment]
Unfortunately this isn't part of the pull request, hence there's no where in code for me to comment; I don't think being able to override the http status code in an error is a good idea. If someone wants to send a different error code, they should create a new error, not override the status code.

**Comment by yunong** [comment]
This is a little bit hard to read through, since not all of the code is part of this PR. I would say 1) remove gulp and use make, and 2) merge and then I'll take a look at the codebase as a whole.

But overall this looks quite good, so kudos!

File: lib/baseClasses/HttpError.js (hunk 5 above)
Comment (DonutEspresso) [inline lib/baseClasses/HttpError.js]: VError does it for us.

File: lib/baseClasses/HttpError.js (hunk 4 above)
Comment (DonutEspresso) [inline lib/baseClasses/HttpError.js]: I'll leave a comment making this clearer, but essentially it boils down to the variadic args. User can opt to pass in `options.message`, but ALSO follow that up with printf style messages that pass through to VError. It's basically this signature:

new Error(options [, sprintf, args...])

In that scenario, VError will generate the printf message for us, and we should prefer that over the `options.message` value. I made a note about it in README, but let me document the code too.

**Comment by DonutEspresso** [comment]
Yeah, I was just thinking the same, I will just merge everything down and we can start discussion around the entire codebase.

I'm maintaining backcompat with existing API today, so overridable status codes exist for both HttpErrors and RestErrors. Would be happy to make that a breaking change for HttpErrors, but I think RestErrors should still be configurable.

**Comment by DonutEspresso** [comment]
BTW, I'm going to merge down the code as is so we can look at it whole. I'll work on the make stuff in a separate PR.

**Pull request merged**